cli/azd/pkg/async/task.go

```diff
@@ -0,0 +1,149 @@
+package async
+
+// Task represents a long running async operation
+type Task[R comparable] struct {
+	isComplete    bool
+	hasResult     bool
+	result        R
+	resultChannel chan R
+	error         error
+}
+
+// Runs the specified taskFn as a go routine
+func (t *Task[R]) Run(taskFn TaskRunFunc[R]) {
+	go func() {
+		context := NewTaskContext(t)
+
+		taskFn(context)
+		t.complete(context.result, context.error)
+	}()
+}
+
+// Checks whether or not the Task has completed
+func (t *Task[R]) IsCompleted() bool {
+	return t.isComplete
+}
+
+// Waits for a result to become available and returns the task value
+func (t *Task[R]) Result() R {
+	if t.hasResult {
+		return t.result
+	}
```

Comment:
What happen if I create a Task and call `Run(fn)` twice? Like:
task := NewTask[Foo]
task.Run( someFunctionThatReturnsFoo )
r := task.Result()  // This should block becaue `hasResult` is false
task.Run( someOTHERFunctionThatReturnsFoo )
r = task.Result()  // Are we returning the previous result from the state instead of blocking again???

I wonder if `Run()` should reset the state (hasResult, isComplete, etc)

Reply:
Updated implementation to force the user to specify the `func` at creation time and will now throw an error if `Run()` is called more than 1 time on the same task object.